README.md
<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/blob/master/src/Banner.PNG?raw=true">
</p>

2D Motion Capture for Unity
-------

[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![PyPI version](https://badge.fury.io/py/unity-2dmocap.svg)](https://badge.fury.io/py/unity-2dmocap)

**2D Motion Capture for Unity** is a python application designed to be a complement of the Unity asset, 2D Motion Capture.
This asset enables an easy construction of 2D sprite animations in Unity by a completely device-less motion capture technique,
and this application plays a role in analyzing a video file to capture the posture of a human in the video. The extraction of the 
posture from the human image is implemented using [OpenCV](https://opencv.org/) and [torchvision](https://pytorch.org/docs/stable/torchvision/index.html).
With this application, you can easily build a sophisticated animation clip for 
your Unity project without any need of special equipments, and once the animation is
generated, you can freely modify the clips the same way you deal with generic Unity animation.

<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/raw/master/src/Demo.gif?raw=true" width=500>
</p>

### Features
* Commandline tool for video analysis and a direct connection to Unity Project
* Simple video editor used for preparation of source video
* Extracts human posture using a deep learning model of keypoint RCNN in torchvision
* Exports posture information to a json file
* Generates Unity animation clip on the editor environment
* Easy save & load system for the sprite rigging on Unity

### Installation
```{commandline}
$ pip install unity-2dmocap
```
To update the application to the latest version, you can run:
```{commandline}
$ pip install --upgrade unity-2dmocap
```
To check if the application is installed properly, you can try the following:
```{commandline}
$ python
```
```{python}
>>> import mocap2d
>>> mocap2d.__version__
```
If it prints out something without an error, the application is installed properly.

### Unity Asset

To use this package, you need a 2D Motion Capture asset, which is a counterpart of 
this package. You can find the asset in the Unity Asset Store.

### Usage

Once you have installed the package, you can use the features through a commandline.

```{commandline}
$ 2dmocap-unity [-h] { detect, send } -f FILE [-g] [-i INTERVAL] [-s] [-o OUT]
```
The commands you can use are "detect" and "send". "detect" will analyze the video file
and connect to the Unity project. "send" will read the pre-generated json file
to the Unity Project without an analysis.

If you are using "detect", the arguments are:
* -h, --help: show the help message
* -f FILE, --file FILE: a path to the video file that you want to analyze
* -g, --gpu: with this argument, the program will run on the gpu instance for a faster analysis
* -i INTERVAL, --interval INTERVAL: a target interval between animation keyframes in seconds
* -s, --show: with this argument, the program will show the resulting skeleton during the analysis
* -o OUT, -output OUT: a directory in which you will save result.json file for the later use

If you are using "send", the arguments are:
* -f FILE, --file FILE: a path to the json file you want to send

### QuickStart

Suppose you want to analyze the video file named "ballet.mp4," and you want to quickly 
analyze it using the gpu of your machine. In addition, you want to have 0.2 seconds of intervals 
between keyframes. Then, the following command will start the program.

```{commandline}
$ 2dmocap-unity detect --file=ballet.mp4 --gpu --interval=0.2 
```

Once you type the command, the program will initiate the analysis as follows.

<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/raw/master/src/usage/1.png?raw=true"
</p>

You can wait until the analysis is over, and if you used the argument -s or --show, 
an external window will show up to show the progress of analysis by drawing a skeleton 
on the original image. After the analysis is complete, the program will address you to connect to 
the Unity project.

<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/raw/master/src/usage/2.png?raw=true"
</p>

To connect to your Unity project, open your project on the Unity editor. If you have have 
successfully imported the "2D Motion Capture" asset, the menu will appear on 
Window/2D Motion Capture.

<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/raw/master/src/usage/3.png?raw=true"
</p>

If you click on the menu, the following editor window will show up. 

<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/raw/master/src/usage/4.png?raw=true"
</p>

As addressed by the program, if you click on "Connect to the Detector" button, the unity will
connect with your python session and receive information about the result of the analysis. Then, 
the window will let you preview the generated animation. If you click on the button "Save Animation Clip", 
the generated animation clip will be saved to the directory you specify.

<p align="center">
    <img src="https://github.com/k2sebeom/unity-2dmocap/raw/master/src/usage/5.png?raw=true"
</p>

Once the animation is generated, you can edit the clip just the same way you
 edit generic unity animation clips.
 
### Community and Feedback

2D Motion Capture for Unity is an open-source project and we encourage and welcome contribution. 
We are waiting for the active participation of game developers, animation artists, and the fans of machine learning technology, so 
if you wish to contribute, please feel free to send pull requests or report issues.

For any further feedback or bug report, please you github's issue template or make a review 
on the Asset Store page. 

Your participation matters a great deal to us.


HISTORY.md
## v0.1.0
* First release of 2D motion capture unity package

## v0.1.1
* Add positional argument of detect / command
* Add feature to save a video analysis result to json file
* Add feature to send a json result directly to unity



2dmocap-unity.py
#!/usr/bin/env python
import argparse
from glob import glob
import json
import warnings
import socket

import cv2

from mocap2d import BoneDetector


if __name__ == '__main__':
    warnings.filterwarnings("ignore", category=UserWarning)
    parser = argparse.ArgumentParser(
        description="Unity extension for 2D motion capture asset"
    )
    parser.add_argument('-f', '--file', dest="file", required=True, nargs=1,
                        help="Path to the video file", type=str)
    parser.add_argument('-g', '--gpu', dest="gpu", action="store_true",
                        help="Run on gpu instance")
    parser.add_argument('-i', '--interval', dest='interval', type=float,
                        help="Interval between key frames in seconds",
                        default=0.2)
    parser.add_argument('-s', '--show', dest='show', action="store_true",
                        help="Show the result of analysis during the process.")
    args = parser.parse_args()

    if not glob(args.file[0]):
        print(f"Video file at {args.file[0]} not found.")
        exit(1)
    print('==' * 50)
    print('''
        #####    #####         ##       ##               #####       
       #######   ##  ###       ####   ####              ####        
            ##   ##    ##      ## ## ## ##             ##             
           ##    ##     ##     ##  ###  ##    ###     ##             
          ##     ##    ##      ##       ##   #   #     ##         ######     ######
        ##       ### ###       ##       ##   #   #      ####     ##    ##    ##   ##
       #######   #####         ##       ##    ###        #####    ####  ##   #####
                                                                             ##
                                                                             ##    by k2sebeom''')
    print("==" * 50 + '\n')
    print("Initializing pose detector...")
    detector = BoneDetector(gpu=args.gpu)
    print()
    print(f"Reading video file at {args.file[0]}")
    video = cv2.VideoCapture(args.file[0])
    fps = video.get(cv2.CAP_PROP_FPS)
    frame_count = int(video.get(cv2.CAP_PROP_FRAME_COUNT))
    interval = args.interval
    fpi = int(interval * fps + 1)
    print(f"{args.file[0]}: FPS = {fps}, FRAME COUNT = {frame_count}")
    print('--' * 50)
    print("Start processing the motion capture")
    bone_list = []
    print(f"[>{'.'*50}] 0%\r", end='')
    for idx in range(frame_count):
        ret, frame = video.read()

        if idx % fpi == 0:
            bones = detector.extract_bone(frame, args.show)
            bones["timestamp"] = round(idx / fps, 3)
            bone_list.append(json.dumps(bones))
            curr = 50 * idx // frame_count
            print(f"[{'=' * curr}>{'.' * (50 - curr)}] "
                  f"{round(100 * idx / frame_count, 2)}%\r", end='')
    print(f"[{'=' * 50}>] 100%")
    cv2.destroyAllWindows()
    video.release()
    print('--' * 50)
    print("Motion Capture complete")
    print('Receive data at Unity by hitting "Connect to Detector" button')

    HOST = "127.0.0.1"
    PORT = 65432

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind((HOST, PORT))
        s.listen()
        conn, addr = s.accept()
        with conn:
            app_name = conn.recv(1024).decode()
            print(f'Connected to Unity project [{app_name}]')
            for bone in bone_list:
                conn.sendall(bone.encode())
                conn.recv(1024)
        print("Transfer Successful")


mocap2d/editor.py
import tkinter as tk
import cv2
import argparse
from glob import glob
import sys
from os import path


global rotate, curr, frame_shape


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('file', help="Path to the video file", type=str)
    args = parser.parse_args()

    if not glob(args.file):
        print(f"Video file at {args.file} not found.")
        exit(1)

    video = cv2.VideoCapture(args.file)
    fps = video.get(cv2.CAP_PROP_FPS)
    frames = []
    while True:
        ret, frame = video.read()
        if not ret:
            break
        frames.append(frame)
    video.release()

    global rotate, curr, frame_shape
    rotate = 0
    curr = 0
    frame_shape = frames[0].shape[:2][::-1]

    def get_image():
        image = frames[curr]
        for _ in range(rotate % 4):
            image = cv2.rotate(image, cv2.ROTATE_90_CLOCKWISE)
        image = cv2.resize(image, frame_shape)
        return image

    def show_image():
        cv2.imshow(title, get_image())
        cv2.waitKey(1)

    window = tk.Tk()
    title = "Simple Video Editor"
    window.title(title)
    window.geometry("500x150")

    def change_point(val):
        global curr
        curr = int(val)
        show_image()

    tk.Label(window, text="Starting point").place(x=5, y=5)
    trim_start = tk.Scale(window, from_=0, to=len(frames) - 1,
                          orient=tk.HORIZONTAL, command=change_point,
                          showvalue=False)
    trim_start.place(x=90, y=5, width=400)

    tk.Label(window, text="Ending point").place(x=5, y=35)
    trim_end = tk.Scale(window, from_=0, to=len(frames) - 1,
                        orient=tk.HORIZONTAL, command=change_point,
                        showvalue=False)
    trim_end.place(x=90, y=35, width=400)
    trim_end.set(len(frames) - 1)

    def rotate_clock():
        global rotate
        rotate += 1
        show_image()

    def rotate_counter():
        global rotate
        rotate -= 1
        show_image()

    btn1 = tk.Button(window, text="CounterClock", command=rotate_counter)
    btn1.place(x=5, y=65)
    btn2 = tk.Button(window, text="ClockWise", command=rotate_clock)
    btn2.place(x=95, y=65)

    tk.Label(window, text="width =").place(x=185, y=65)
    w = tk.Entry(window)
    w.place(x=235, y=65, width=40)
    tk.Label(window, text="height =").place(x=285, y=65)
    h = tk.Entry(window)
    h.place(x=345, y=65, width=40)

    def resize():
        global frame_shape
        if w.get() and h.get() and int(w.get()) > 0 and int(h.get()) > 0:
            frame_shape = (int(w.get()), int(h.get()))
            show_image()

    tk.Button(window, text="Resize", command=resize).place(
        x=395, y=60, width=70)

    def save_video():
        global curr
        if int(trim_start.get()) > int(trim_end.get()):
            print("Invalid Trim points")

        fourcc = cv2.VideoWriter_fourcc(*'mp4v')
        filename = path.basename(args.file[0]).split('.')[0] + '.mp4'
        new_path = path.join(path.dirname(args.file[0]), 'new-' + filename)
        vid_shape = get_image().shape[:2][::-1]
        out = cv2.VideoWriter(new_path, fourcc, fps, vid_shape)
        for idx in range(int(trim_start.get()), int(trim_end.get()) + 1):
            curr = idx
            show_image()
            out.write(get_image())
        out.release()
        print(f"Generated Video at {new_path}")

    tk.Button(window, text="Save", command=save_video).place(
        x=5, y=100, width=490, height=40)

    show_image()

    window.mainloop()


mocap2d/mocap2d.py
import torch
import cv2
import numpy as np

from torchvision.models.detection import keypointrcnn_resnet50_fpn as keypoint


class BoneDetector:
    __BONES = [
        (17, 0), (17, 5), (5, 7), (7, 9), (17, 6), (6, 8), (8, 10), (17, 18),
        (18, 11), (11, 13), (13, 15), (18, 12), (12, 14), (14, 16)
    ]

    def __init__(self, gpu=False):
        self._model = keypoint(pretrained=True)
        self._device = None
        if gpu:
            print("Detecting gpu device...")
            self._device = torch.device("cuda:0")
            if self._device is None:
                raise Exception("No GPU instance detected!!")
            self._model.to(self._device)
        self._model.eval()
        print("Pose detector initialized.")

    def _image_to_tensor(self, image):
        img = np.transpose(image / 255, (2, 0, 1))
        img = torch.from_numpy(img).float()
        if self._device:
            img = img.to(self._device)
        return img

    def extract_bone(self, image, show=False):
        im_tensor = self._image_to_tensor(image)
        result = self._model([im_tensor])[0]
        points, scores = self._find_human(result)
        if points is None:
            return None
        if show:
            self._draw_human(image, points)
        skeleton = {}
        for idx, keys in enumerate(self.__BONES):
            k1, k2 = keys
            skeleton[f"bone{idx + 1}"] = self._get_bone_vector(
                points[k1], points[k2])
        return skeleton, points[17]

    def _draw_human(self, image, points):
        for pt1, pt2 in self.__BONES:
            image = cv2.line(image, tuple(points[pt1][:2]),
                             tuple(points[pt2][:2]), (255, 0, 0), 3)
        cv2.imshow("Result", image)
        cv2.waitKey(1)

    @staticmethod
    def _get_bone_vector(key1, key2):
        return int(key2[0] - key1[0]), int(key1[1] - key2[1])

    def _find_human(self, raw_result):
        scores = raw_result["scores"].data.cpu().numpy()
        if len(scores) == 0:
            return None, None
        human_idx = np.argmax(scores)
        points = raw_result["keypoints"][human_idx].data.cpu().numpy()
        scores = raw_result["keypoints_scores"][human_idx].data.cpu().numpy()
        mid_top = self._mid_point(points[5], points[6])
        mid_bot = self._mid_point(points[11], points[12])
        points = np.append(points, [mid_top, mid_bot], axis=0).astype(np.int32)
        scores = np.append(
            scores, [scores[5] + scores[6], scores[11] + scores[12]]
        )
        return points, scores

    @staticmethod
    def _mid_point(point1, point2):
        return (point1[0] + point2[0]) // 2, (point1[1] + point2[1]) // 2, 1


mocap2d/detect.py
import argparse
from glob import glob
import json
import warnings
import socket
from os import path

import cv2

from mocap2d import BoneDetector


def detect(args):
    if args.out and not path.isdir(args.out):
        print(f"Output path {args.out} is not a directory")
        exit(2)

    if not glob(args.file[0]):
        print(f"Video file at {args.file[0]} not found.")
        exit(1)

    print("Initializing pose detector...")
    detector = BoneDetector(gpu=args.gpu)
    print()
    print(f"Reading video file at {args.file[0]}")
    video = cv2.VideoCapture(args.file[0])
    fps = video.get(cv2.CAP_PROP_FPS)
    frame_count = int(video.get(cv2.CAP_PROP_FRAME_COUNT))
    interval = args.interval
    fpi = int(interval * fps + 1)
    print(f"{args.file[0]}: FPS = {fps}, FRAME COUNT = {frame_count}")
    print('--' * 50)
    print("Start processing the motion capture")
    bone_list = []
    print(f"[>{'.' * 50}] 0%\r", end='')

    pos0 = None
    for idx in range(frame_count):
        ret, frame = video.read()

        if idx % fpi == 0:
            bones, pos = detector.extract_bone(frame, args.show)
            if pos0 is None:
                pos0 = pos
            print(f"{pos[0] - pos0[0]}, {pos[1] - pos0[1]}")
            bones["timestamp"] = round(idx / fps, 3)
            bone_list.append(json.dumps(bones))
            curr = 50 * idx // frame_count
            print(f"[{'=' * curr}>{'.' * (50 - curr)}] "
                  f"{round(100 * idx / frame_count, 2)}%\r", end='')
    print(f"[{'=' * 50}>] 100%     ")
    cv2.destroyAllWindows()
    video.release()

    if args.out:
        with open(path.join(args.out, "result.json"), 'w') as out_file:
            out_file.writelines([bone + '\n' for bone in bone_list])

    print('--' * 50)
    print("Motion Capture complete")
    return bone_list


def send(args):
    if '.json' not in args.file[0]:
        print("Please provide the json file with the pose data")
        exit(4)

    print(f"Retrieving data from {args.file[0]}")
    bone_list = []
    with open(args.file[0], 'r') as json_file:
        line = json_file.readline()
        while line:
            bone_list.append(line)
            line = json_file.readline()
    print("Pose data is ready\n")
    return bone_list


def main():
    warnings.filterwarnings("ignore", category=UserWarning)
    parser = argparse.ArgumentParser(
        description="Unity extension for 2D motion capture asset"
    )
    parser.add_argument('command', metavar='{ detect, send }',
                        help="detect: analyze video file and send / "
                             "send: send json result without analyzing",
                        type=str)

    parser.add_argument('-f', '--file', dest="file", required=True, nargs=1,
                        help="Path to the video file", type=str)
    parser.add_argument('-g', '--gpu', dest="gpu", action="store_true",
                        help="Run on gpu instance")
    parser.add_argument('-i', '--interval', dest='interval', type=float,
                        help="Interval between key frames in seconds",
                        default=0.2)
    parser.add_argument('-s', '--show', dest='show', action="store_true",
                        help="Show the result of analysis during the process.")
    parser.add_argument('-o', '--output', dest='out', type=str,
                        help="Directory to save the result in a json format")
    args = parser.parse_args()

    print('==' * 50)
    print('''
        #####    #####         ##       ##               #####       
       #######   ##  ###       ####   ####              ####        
            ##   ##    ##      ## ## ## ##             ##             
           ##    ##     ##     ##  ###  ##    ###     ##             
          ##     ##    ##      ##       ##   #   #     ##         ######     ######
        ##       ### ###       ##       ##   #   #      ####     ##    ##    ##   ##
       #######   #####         ##       ##    ###        #####    ####  ##   #####
                                                                             ##
                                                                             ##    by k2sebeom''')
    print("==" * 50 + '\n')

    bone_list = []
    if args.command == "detect":
        bone_list = detect(args)
    elif args.command == "send":
        bone_list = send(args)
    else:
        print("Please enter a valid commad of either 'send' or 'detect'")
        exit(3)

    print('Receive data at Unity by hitting "Connect to Detector" button')

    host = "127.0.0.1"
    port = 65432

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind((host, port))
        s.listen()
        conn, addr = s.accept()
        with conn:
            app_name = conn.recv(1024).decode()
            print(f'Connected to Unity project [{app_name}]')
            for bone in bone_list:
                conn.sendall(bone.encode())
                conn.recv(1024)
        print("Transfer Successful")


if __name__ == '__main__':
    main()
